app/topics_handler.go

```diff
@@ -61,87 +61,84 @@ func (th *TopicsHandler) requestTopicWorkers(ctx sdk.Context, topic emissionstyp
-	// iterate over all the worker nonces to find if this is unfulfilled
-	for _, nonce := range sortedWorkerNonces {
-		nonceCopy := nonce
-		Logger(ctx).Debug(fmt.Sprintf("Current Worker block height has been found unfulfilled, requesting inferences %v", nonceCopy))
-		go generateInferencesRequest(ctx, topic.InferenceLogic, topic.InferenceMethod, topic.DefaultArg, topic.Id, topic.AllowNegative, *nonceCopy)
+	if len(workerNonces.Nonces) < 1 {
+		Logger(ctx).Warn("No unfulfilled worker nonces found for topic %d", topic.Id)
+		return
 	}
+	nonceCopy := workerNonces.Nonces[0]
+
+	Logger(ctx).Debug(fmt.Sprintf("Unfulfilled worker nonce found, requesting inferences %v", nonceCopy))
+	go generateInferencesRequest(ctx, topic.InferenceLogic, topic.InferenceMethod, topic.DefaultArg, topic.Id, topic.AllowNegative, *nonceCopy)
 }
 
 func (th *TopicsHandler) requestTopicReputers(ctx sdk.Context, topic emissionstypes.Topic) {
-	currentBlockHeight := ctx.BlockHeight()
 	Logger(ctx).Debug(fmt.Sprintf("Triggering Losses cadence met for topic: %v metadata: %s default arg: %s \n",
 		topic.Id, topic.Metadata, topic.DefaultArg))
 	reputerNonces, err := th.emissionsKeeper.GetUnfulfilledReputerNonces(ctx, topic.Id)
 	if err != nil {
 		Logger(ctx).Error("Error getting reputer nonces: " + err.Error())
 		return
 	}
-	// No filtering - reputation of previous rounds can still be retried if work has been done.
-	maxRetriesToFulfilNoncesReputer := emissionstypes.DefaultParams().MaxRetriesToFulfilNoncesReputer
-	emissionsParams, err := th.emissionsKeeper.GetParams(ctx)
-	if err != nil {
-		Logger(ctx).Warn(fmt.Sprintf("Error getting max num of retries to fulfil nonces for worker requests (using default), err: %s", err.Error()))
-	} else {
-		maxRetriesToFulfilNoncesReputer = emissionsParams.MaxRetriesToFulfilNoncesReputer
+
+	if len(reputerNonces.Nonces) < 1 {
+		Logger(ctx).Warn("No unfulfilled reputer nonces found for topic %d", topic.Id)
+		return
 	}
-	topNReputerNonces := synth.SelectTopNReputerNonces(&reputerNonces, int(maxRetriesToFulfilNoncesReputer), currentBlockHeight, topic.GroundTruthLag, topic.EpochLength)
-	Logger(ctx).Warn(fmt.Sprintf("Iterating Top N Reputer Nonces: %v", len(topNReputerNonces)))
 
 	lastCommit, err := th.emissionsKeeper.GetTopicLastCommit(ctx, topic.Id, emissionstypes.ActorType_REPUTER)
 	if err != nil {
 		Logger(ctx).Warn("Error getting reputer last commit: "+err.Error(), ", first reputer commit?")
 	}
-	// iterate over all the reputer nonces to find if this is unfulfilled
-	for _, nonce := range topNReputerNonces {
-		nonceCopy := nonce
-		// Get previous losses from keeper, default: reputer - epochLength
-		lastReputerCommitBlockHeight := nonceCopy.ReputerNonce.BlockHeight - topic.EpochLength
-		if lastCommit.Nonce != nil {
-			Logger(ctx).Debug(fmt.Sprintf("Reputer last commit found, setting: %v", lastCommit))
-			lastReputerCommitBlockHeight = lastCommit.Nonce.BlockHeight
-		}
+	nonceCopy := reputerNonces.Nonces[0]
```

Comment:
I think this one is trickier: because reputer nonces legitimately accumulate during ground_truth_lag, before they get to be processed. 
But they can also accumulate because of reputation rounds missed (just like inference rounds missed above).

So this code would also get stuck with the earliest reputation attempt (if new nonces are appended at end), 
So we should get those eligible first (ie those where gt is revealed already), and get the last one of those.

... or do like we were doing: do a number of them and retry, giving priority to new ones, and retrying those unfulfilled ones until they get pruned.

Reply:
we could send out up to a batch per topic each time, then we're guaranteed to eventually catch up so long as that batch size >1

Reply:
maybe we also need a new keyset/map that tracks if the unfulfilled nonce was attempted yet